auth.go

```diff
@@ -0,0 +1,202 @@
+func NewAutoAuth(login string, secret string) Authorizer {
+	fmap := make(map[string]AuthFactory)
+	az := &authorizer{fmap, sync.Mutex{}, &nullAuth{}}
+
+	az.AddAuthenticator("basic", func(rq *http.Request, rs *http.Response, method, path string) (auth Authenticator, err error) {
+		return &BasicAuth{login, secret}, nil
+	})
+
+	az.AddAuthenticator("digest", func(rq *http.Request, rs *http.Response, method, path string) (auth Authenticator, err error) {
+		return &DigestAuth{login, secret, digestParts(rs)}, nil
+	})
+
+	return az
+}
+
+// NewAuthenticator creates an Authenticator (Shim) per request
+func (a *authorizer) NewAuthenticator(body io.Reader) (Authenticator, io.Reader) {
+	var retryBuf io.Reader = body
+	if body != nil {
+		// If the authorization fails, we will need to restart reading
+		// from the passed body stream.
+		// When body is seekable, use seek to reset the streams
+		// cursor to the start.
+		// Otherwise, copy the stream into a buffer while uploading
+		// and use the buffers content on retry.
+		if _, ok := retryBuf.(io.Seeker); ok {
+			body = io.NopCloser(body)
+		} else {
+			buff := &bytes.Buffer{}
+			retryBuf = buff
+			body = io.TeeReader(body, buff)
+		}
+	}
+	a.defAuthMux.Lock()
+	defAuth := a.defAuth.Clone()
+	a.defAuthMux.Unlock()
+
+	return &authShim{a.factory, retryBuf, defAuth}, body
+}
+
+func (a *authorizer) AddAuthenticator(key string, fn AuthFactory) {
+	a.factories[key] = fn
+}
+
+// factory creates an Authenticator instance based on the WWW-Authenticate header
+func (a *authorizer) factory(rq *http.Request, rs *http.Response, method, path string) (auth Authenticator, err error) {
+	header := strings.ToLower(rs.Header.Get("Www-Authenticate"))
+	for k, fn := range a.factories {
+		if strings.Contains(header, k) {
+			if auth, err = fn(rq, rs, method, path); err != nil {
+				return
+			}
+			break
+		}
+	}
+	if auth == nil {
+		return nil, newPathError("NoAuthenticator", path, rs.StatusCode)
+	}
+
+	a.defAuthMux.Lock()
+	a.defAuth = auth
+	a.defAuthMux.Unlock()
+
+	return auth, nil
+}
+
+// Authorize the current request
+func (s *authShim) Authorize(c *http.Client, rq *http.Request, method string, path string) error {
+	if err := s.auth.Authorize(c, rq, method, path); err != nil {
+		return err
+	}
+	body := s.body
+	rq.GetBody = func() (io.ReadCloser, error) {
+		if body != nil {
+			if sk, ok := body.(io.Seeker); ok {
+				if _, err := sk.Seek(0, io.SeekStart); err != nil {
+					return nil, err
+				}
+			}
+			return io.NopCloser(body), nil
+		}
+		return nil, nil
+	}
+	return nil
+}
+
+// Verify checks for authentication issues and may trigger a re-authentication.
+// Catches AlgoChangedErr to update the current Authenticator
+func (s *authShim) Verify(rq *http.Request, rs *http.Response, method string, path string) (reauth bool, err error) {
+	reauth, err = s.auth.Verify(rq, rs, method, path)
+	if err != nil {
+		if _, ok := err.(AlgoChangedErr); ok {
```

Comment:
Isn't it better if this error would be a sentinel error at package level (`var AlgoChangedErr = errors.New("AuthChangedErr")`) and the check here done with `errors.Is(err, AlgoChangedErr)` ?

Reply:
`errors.go`seems to be a better place?